app/abci.go

```diff
@@ -341,6 +377,167 @@ func (app *HeimdallApp) VerifyVoteExtensionHandler() sdk.VerifyVoteExtensionHand
+		return err
+	}
+
+	var lastMilestone *milestoneTypes.Milestone
+
+	if hasMilestone {
+		lastMilestone, err = app.MilestoneKeeper.GetLastMilestone(ctx)
+		if err != nil {
+			logger.Error("Error occurred while fetching the last milestone", "error", err)
+			return err
+		}
+	}
+
+	lastMilestoneBlock, err := app.MilestoneKeeper.GetLastMilestoneBlock(ctx)
+	if err != nil {
+		logger.Error("Error occurred while fetching the last milestone block", "error", err)
+		return err
+	}
+
+	if lastMilestoneBlock == 0 {
+		lastMilestoneBlock = uint64(ctx.BlockHeight())
+	}
+
+	diff := ctx.BlockHeight() - int64(lastMilestoneBlock)
+
+	if lastMilestone != nil && lastMilestoneBlock != 0 && diff > ChangeProducerThreshold && helper.IsVeblop(lastMilestone.EndBlock+1) {
+		logger.Info("Block finalization time is greater than change producer threshold, creating a new veblop span", "lastMilestone", lastMilestone, "lastMilestoneBlock", lastMilestoneBlock, "diff", diff, "currentBlock", ctx.BlockHeight())
+
+		addSpanCtx, spanCache := app.cacheTxContext(ctx)
+
+		latestActiveProducer, err := app.BorKeeper.GetLatestActiveProducer(ctx)
+		if err != nil {
+			logger.Error("Error occurred while getting latest active producer", "error", err)
+			return err
+		}
+
+		lastSpan, err := app.BorKeeper.GetLastSpan(ctx)
+		if err != nil {
+			logger.Error("Error occurred while getting last span", "error", err)
+			return err
+		}
+
+		params, err := app.BorKeeper.GetParams(ctx)
+		if err != nil {
+			logger.Error("Error occurred while getting bor params", "error", err)
+			return err
+		}
+
+		endBlock := lastSpan.EndBlock
+
+		for endBlock-lastMilestone.EndBlock > 2*params.SpanDuration {
+			endBlock -= params.SpanDuration
+		}
+
+		if endBlock <= lastMilestone.EndBlock {
+			endBlock = lastSpan.EndBlock
+			for endBlock <= lastMilestone.EndBlock {
+				endBlock += params.SpanDuration
+			}
+		}
+
+		currentProducer, err := app.BorKeeper.FindCurrentProducerID(ctx, lastMilestone.EndBlock+1)
+		if err != nil {
+			logger.Error("Error occurred while finding current producer", "error", err)
+			return err
+		}
+
+		latestFailedProducer, err := app.BorKeeper.GetLatestFailedProducer(ctx)
+		if err != nil {
+			logger.Error("Error occurred while getting latest failed producer", "error", err)
+			return err
+		}
+
+		for producerID := range latestFailedProducer {
+			delete(latestActiveProducer, producerID)
+		}
+
+		delete(latestActiveProducer, currentProducer)
+
+		err = app.BorKeeper.AddNewVeblopSpan(addSpanCtx, currentProducer, lastMilestone.EndBlock+1, endBlock, lastMilestone.BorChainId, latestActiveProducer)
+		if err != nil {
+			logger.Warn("Error occurred while adding new veblop span", "error", err)
+		} else {
+			// update the last milestone block to a future block height to avoid immediately rotating the span in the next block
+			err = app.MilestoneKeeper.SetLastMilestoneBlock(addSpanCtx, uint64(ctx.BlockHeight())+SpanRotationBuffer)
```

Comment:
So incase next producer is not ready, we can have ~2 minutes down time before rotating again?

Reply:
I was assuming 60 blocks is 1 min. It might be too long, and we can probably reduce it once the point above (span <> milestone conflict) is addressed.